### DISCUSSION TITLE
Interpolation and constraints

### DISCUSSION THREAD

**Discussion by micycle1** [gwlucastrig/Tinfour#68] [Q&A]
I have a few points to make about how constraints interact with interpolation.

### Without constraints
Interpolation of a triangulated vertex set without constraints.

This is pretty much expected: a interpolation field across the convex hull of the vertices.
![image](https://user-images.githubusercontent.com/9304234/111921916-0db73900-8a8f-11eb-8886-e38db0168521.png)

### With constraints
A screenshot of interpolation applied to a tin with a _PolygonConstraint_ applied (based on the vertices).

My first observation is that interpolators interpolate across **unconstrained** regions — would it be more appropriate if they return _NaN_ for such areas instead?

I notice the intended colors aren't really there anymore. There either averaged away somehow or I suspect this is because it's ignoring z values of vertices on the constrained edges? If so, that's rather problematic as that's where all the vertices are!
![image](https://user-images.githubusercontent.com/9304234/111921880-e496a880-8a8e-11eb-8c13-23f091853e57.png)

### Desired

Desirable: interpolator returns _NaN_ for coordinates **outside** the constrained region. Vertices that lie on the polygon constraint (the edge) contribute properly to the constrained region's interpolated values.

![image](https://user-images.githubusercontent.com/9304234/111921927-245d9000-8a8f-11eb-9996-b761c48ad875.png)

**Comment by gwlucastrig** [comment]
I'm going to have to think a bit about how to better support your objectives.  But for now, I will try to clarify the way area constraints work in Tinfour.  Tinfour grew out of geophysical applications, and some of the design choices reflect that origin... I think constraints are one of them.

In Tinfour, a constraint is really just a way of partitioning the coordinate plane into separate areas. Think of water/land or even the case of a roadbed (which marks a 1st derivative discontinuity in a surface). So an interpolation is valid on either side of the partition, just as long as it doesn't cross the partition.  And, if properly implemented, an interpolator will not cross a constraint (though it occurs to me that the recently introduced the Inverse Distance Weighting interpolator doesn't honor that requirement).

To address your problem, you could either implement a pre-test or extend the interpolator class to:

1. Call getNavigator() on your IncrementalTin instance to get an IIncrementatlTinNavigator instance.
2. Call the navigator to getNeighborEdge() or getNearestEdge() for your interpolation coordinates to get a neighboring edge
3. Call IQuadEdge.getConstraintIndex() on the edge to find out which constraint it is associated with
4. Maybe access IQuadEdge isConstraintRegionInterior() or IQuadEdge.isConstraintRegionBorder() to further refine the meaning of the edge.
5. The above steps will identify the constraint index for the area containing the interpolation point. Giving that information, your code can decide to move forward with interpolation (if the point is inside a valid area) or return a NaN.

You might find it works better to try using the navigator's getContainingTriangle() to clarify the assignment of the local geometry.

If you'd like I can try to put together some demo code illustrating the concept described above.

Also, just out of curiosity, which interpolator are you interested in using?

**Reply by micycle1** [reply]
I've been using the `NaturalNeighbor` and `TriangularFacet` interpolators.

How does what's happening in image 2 (interpolating a tin with constraints) fit in with what you've mentioned here? Should the output be different from image 1?

**Comment by gwlucastrig** [comment]
Revision to above notes...   Actually, I would recommend calling the navigator's getContainingTriangle() method.  And then calling the SimpleTriangles getContainingRegion() call.   The Javadoc in the code for getContainingRegion()  warns about a potential problem, but that Javadoc is out-of-date.  I fixed that problem for   [Issue 47](https://github.com/gwlucastrig/Tinfour/issues/47), but failed to update the Javadoc.

**Comment by gwlucastrig** [comment]
Looking at this posting again, I see that I haven't really addressed your question.  I will need some time to think about this problem a bit more.  I'll try to look at it this evening.
